Better page for Response.type (#39541)
mdn/content: files/en-us/web/api/response/type/index.md

{{APIRef("Fetch API")}}{{AvailableInWorkers}}

> The **`type`** read-only property of the {{domxref("Response")}} interface contains the type of the response.

Author: ```suggestion
The **`type`** read-only property of the {{domxref("Response")}} interface contains the type of the response. The type determines whether scripts are able to access the response body and headers.
```

## Value

A string, which may be any of the following values:

- `basic`

  - : This applies in any of the following cases:

    - The request is same-origin.
    - The requested URL's scheme is [`data:`](/en-US/docs/Web/URI/Reference/Schemes/data).
    - The request's {{domxref("Request.mode", "mode")}} is `navigate` or `websocket`.

>     All response headers are exposed except {{httpheader("Set-Cookie")}}.
> 
> - `cors`
>   - : The request was cross-origin and was successfully processed using [CORS](/en-US/docs/Web/HTTP/Guides/CORS). Only {{glossary("CORS-safelisted response header", "CORS-safelisted response headers")}} are exposed in the response.

Author: ```suggestion
    With this type, all response headers are exposed except {{httpheader("Set-Cookie")}}.

- `cors`
  - : The request was cross-origin and was successfully processed using [CORS](/en-US/docs/Web/HTTP/Guides/CORS). With this type, only {{glossary("CORS-safelisted response header", "CORS-safelisted response headers")}} are exposed.
```

- `opaque`

  - : A response to a cross-origin request whose {{domxref("Request.mode", "mode")}} was set to `no-cors`. The {{domxref("Response.status", "status")}} property is set to `0`, {{domxref("Response.body", "body")}} is `null`, headers are empty and immutable.

- `opaqueredirect`
  - : A response to a request whose {{domxref("Request.redirect", "redirect")}} option was set to `manual`, and which was redirected by the server. The {{domxref("Response.status", "status")}} property is set to `0`, {{domxref("Response.body", "body")}} is `null`, headers are empty and immutable.

## Examples

### A basic response

The following same-origin request will return a `basic` response:

```js
const response = await fetch("flowers.jpg");

console.log(response.type);
// "basic"
```

### A CORS response

Assuming `https://example.org` is not the requester's origin, and that the server responds with the appropriate CORS headers, this request will return a `cors` response:

```js
const response = await fetch("https://example.org/flowers.jpg");

console.log(response.type);
// "cors"
```

### An opaque response

> The following response is made with the [`no-cors`](/en-US/docs/Web/API/Request/mode#no-cors) option, so the reponse's type is `opaque`:

Author: ```suggestion
The following request is made with the [`no-cors`](/en-US/docs/Web/API/Request/mode#no-cors) option, so it returns an `opaque` response:
```

```js
const response = await fetch("https://example.org/flowers.jpg", {
  mode: "no-cors",
});

console.log(response.type);
// "opaque"
console.log(response.body);
// null
console.log(response.status);
// 0
```

### An error response

> The following response is constructed using {{domxref("Response.error_static", "Response.error()")}} and is of type `error`:

Author: ```suggestion
The following code uses {{domxref("Response.error_static", "Response.error()")}} to create an `error` response:
```
